### PR TITLE
executor: support SELECT FOR UPDATE for BatchPointGet

### PR DESCRIPTION
**Pull request by coocood**
### What problem does this PR solve? 
improve performance

### What is changed and how it works?
Make BatchPointGet support select for update statements, 

### Check List 

Tests 

 - Unit test

Release note

 - support SELECT FOR UPDATE for BatchPointGet

### CODE REVIEW

File: executor/batch_point_get.go [hunk 1]
```diff
@@ -77,26 +79,38 @@ func (e *BatchPointGetExec) Next(ctx context.Context, req *chunk.Chunk) error {
...
 		return err
 	}
-	var reader kv.Snapshot
-	if !txn.Valid() {
+	var batchGetter kv.BatchGetter
+	if txn.Valid() {
 		// In restricted SQL, !txn.Valid() may happen, we doesn't call PrepareTxnCtx but
 		// call Next() and the code runs here.
-		reader = e.snapshot
+		if e.snapshotTS == e.startTS {
```
Comment (cfzjywxk) [inline executor/batch_point_get.go]: For read-consistency mode,  `BatchPointGetExec.startTS` may be the same with `forUpdateTS`,  here using txn as batchGetter will read wrong data?

File: executor/batch_point_get.go [hunk 2]
```diff
@@ -155,9 +169,47 @@ func (e *BatchPointGetExec) initialize(ctx context.Context) error {
 	}
 
 	// Fetch all values.
-	values, err1 := reader.BatchGet(ctx, keys)
-	if err1 != nil {
-		return err1
+	var values map[string][]byte
+	if e.lock {
```
Comment (cfzjywxk) [inline executor/batch_point_get.go]: The `initialize` function seems a bit long, better  put lock into another function like point get `LockKeysIfNeeded`

File: executor/batch_point_get_test.go [hunk 3]
```diff
@@ -111,11 +111,20 @@ func (s *testBatchPointGetSuite) TestBatchPointGetInTxn(c *C) {
 	tk.MustExec("begin")
 	tk.MustExec("insert into t values (4, 'name')")
 	tk.MustQuery("select * from t where id in (4)").Check(testkit.Rows("4 name"))
+	tk.MustQuery("select * from t where id in (4) for update").Check(testkit.Rows("4 name"))
 	tk.MustExec("rollback")
 
 	// Pessimistic transaction doesn't use BatchPointGet! Make sure this work.
```
Comment (cfzjywxk) [inline executor/batch_point_get_test.go]: Seems this comment could be removed

File: executor/batch_point_get.go (hunk 1 above)
Comment (coocood) [reply executor/batch_point_get.go]: RC would update the snapshot, so this doesn't affect correctness, but the logic here is useless even for REPEATABLE-READ level, the `e.startTS` is not the real startTS of the transaction we should rename it in the future PR.

**Comment by coocood** [comment]
@cfzjywxk PTAL

**Comment by cfzjywxk** [comment]
LGTM

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**